modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheAffinitySharedManager.java

```diff
@@ -2920,6 +2921,46 @@ private CacheGroupNoAffOrFilteredHolder createHolder(
         return new CacheGroupNoAffOrFilteredHolder(ccfg.getRebalanceMode() != NONE, cctx, aff, initAff);
     }
 
+    /**
+     * Prints {@code waitInfo} to provided {@code log}.
+     *
+     * @param log Logger to print to.
+     */
+    public void printWaitInfo(IgniteLogger log) {
+        if (waitInfo == null || waitInfo.empty())
+            return;
+
+        StringBuilder sb = new StringBuilder();
+
+        sb.append("Current affinity assignment is not ideal, it is waiting for cache: [");
+
+        waitInfo.assignments.forEach((grpId, partsMap) -> {
+            sb.append("grp=[grpId=").append(grpId).append(",nodes=[");
+
+            Map<UUID/* ClusterNode id */, Integer/* Number of partitions for this UUID */> nodeMap = new HashMap<>();
+
+            partsMap.forEach((partId, nodes) -> {
+                nodes.forEach((node -> {
```

Comment:
Ensure all the collections cannot throw CME.

Reply:
Now I'm using a copy of the partition map for string construction. Please confirm that it's enough